design/secretless-e2e-tests-dnd.md

```diff
@@ -0,0 +1,256 @@
+
+The goal of this particular effort is to improve the process of creating and maintaining Secretless end-to-end tests. It is inspired by the immediate need to add a test case for Conjur OSS. Our current test cases assume success implicitly as running through some number of steps without encountering failure. WHAT we test requires some thought and improvement, however we deffer this to a later effort while keeping in mind that this addressing this technical debt will become simpler as a result of an improved process.
+
+The improvements in process are expected to allow developers on the Secretless project to focus efforts on writing integration test cases with the assurance that some robust other thing will handle
+1. dependency service lifecycle management
+2. running of the tests
+3. measuring and reporting the outcomes of the tests
+
+# Technical Objectives
+
+_At a high-level, what are the technical requirements that this design needs to satisfy?  How can it meet the acceptance criteria presented in the Feature Proposal?_
+
+- [ ] Tests run in a realistic environment with DAP/OSS as the source of truth for secret data
+- [ ] Tests run on developer machine ?
+- [ ] Tests run in Kubernetes and OpenShift
+- [ ] Tests run against relevant versions of DAP (v10+), and Conjur OSS
+- [ ] Tests run against PostgreSQL and MySQL
+- [ ] Tests validate that a demo application is able to run as expected while making database connections via Secretless
+- [ ] The test suite does not need to run with every PR of Secretless, but should run daily and we should be able to run it against a local build of the Secretless image to prep for a release
+
+## Out of Scope
+
+1. Creating sophisticated implementations of service lifecycle management will not be part of this effort e.g. pooling pre-existing dependency resources to streamline load times
+2. Being exhaustive in testing all the possible versions of services like Conjur. Though this would be thorough this would be time consuming and costly. In the future we might consider having test scenarios that are less costly and less representative of production but close enough to it to be of value. An example is running a branch follower against a stable master.
+3. juxtaposer will eventually need to be automated. This is not addressed as part of this effort
+|
+
+# Experience
+
+## Assumptions
+
+We assume that the experience of running automated tests given configured dependency resources (such as Conjur and target services) is decent enough at this point in time to not warrant consideration in the improvements suggested by this document. This means  a user (a developer or Jenkins) is able to run integration tests in an automated fashion and get meaningful feedback on the test outcome.
+
+Moving the lifecycle management of dependency resources behind an opaque API is a task that can be achieved with minimal effort.
```

Comment:
this seems like a big assumption - is it true? is it part of the purpose of the design doc to define what doing this will entail, which will help us discover whether this is true?